### PR TITLE
If the user has "fixendofline" set, append a blank line to preserve the EOF newline

### LINKED ISSUE #301: didOpen sends document text without final newline
When opening a document, `vim-lsp` currently sends the document without the final newline. This causes a problem when calling `:LspFormatDocument` with formatters like `black` and `prettier` because they will add a final newline if it is missing which signals to the language server that the file contents have changed and then adds a modification to the undo stack that doesn't have anything to undo. To get around this, `vim-lsp` should send the contents of the document to the server with a newline appended if `fixendofline` is set. The following seems to work fairly well (and is similar to what `LanguageClient-neovim` does):

**autoload/lsp.vim**
```
function! s:get_text_document_text(buf, server_name) abort
	let l:file_content = s:get_last_file_content(a:buf, a:server_name)
	if len(l:file_content) > 0 && l:file_content[-1] !=# '' && &fixendofline
		let l:file_content += ['']
	endif
    return join(l:file_content, "\n")
endfunction
```

### PR DESCRIPTION
**Pull request by bryanforbes**
Fixes #301

### CODE REVIEW

**Comment by prabirshrestha** [comment]
@greghensley Can you try this. This uses a bit different implementation compared to this PR https://github.com/prabirshrestha/vim-lsp/pull/267. But given that LSP doesn't care about line ending I think this would work too.

File: autoload/lsp.vim [hunk 1]
```diff
@@ -235,6 +235,14 @@ function! s:on_text_document_did_close() abort
     call lsp#log('s:on_text_document_did_close()', l:buf)
 endfunction
 
+function! s:get_buffer_lines(buf) abort
```
Comment (prabirshrestha) [inline autoload/lsp.vim]: Can we move this to util and make it public but start with `_` and have tests for it. We have regressed several times on this so having tests would help here. `lsp#utils#buffer#_get_lines(buf)`

**Comment by bryanforbes** [comment]
> Could you add tests for this.

Absolutely, but how do I run the tests? I don't see any information about it.

File: autoload/lsp.vim (hunk 1 above)
Comment (bryanforbes) [reply autoload/lsp.vim]: Yep! I'll make the change later tonight.

**Comment by bryanforbes** [comment]
I just took a look at @greghensley's PR and I think it would be good to incorporate it into the utility function @prabirshrestha would like me to split out. If it's ok with @greghensley, I'll combine our efforts into this PR and write a test.

File: autoload/lsp.vim [hunk 2]
```diff
@@ -235,6 +235,14 @@ function! s:on_text_document_did_close() abort
...
+	let l:lines = getbufline(a:buf, 1, '$')
+	if len(l:lines) > 0 && l:lines[-1] !=# '' && &fixendofline
+		let l:lines += ['']
+	endif
+	return l:lines
+endfunction
```
Comment (mattn) [inline autoload/lsp.vim]: Please use 4 spaces instead of tab character.

**Comment by prabirshrestha** [comment]
Should be ok to combine.

We do need to document how to run tests. In the mean time you can follow instruction in docs at 
https://github.com/thinca/vim-themis and run `themis --reporter spec`.

**Comment by bryanforbes** [comment]
Ok, I've split the utility function out; improved the detection based on the Vim 7.3 (which didn't have `fixendofline`) and 8.0 documentation for `endofline`, `binary`, and (in the case of 8.0) `fixendofline`; and added tests for all of the cases.

**Comment by prabirshrestha** [comment]
Merged. Thanks for the fix as well as tests. Test will help us to never regress this again 😄

**Pull request merged**